heron/scheduler-core/src/java/com/twitter/heron/scheduler/UpdateTopologyManager.java

```diff
@@ -0,0 +1,249 @@
+    logFine("Deleted existing Topology: %s", stateManager.deleteTopology(topologyName));
+    logFine("Set new Topology: %s", stateManager.setTopology(updatedTopology, topologyName));
+
+    // update packing plan to trigger the scaling event
+    logFine("Deleted existing packing plan: %s", stateManager.deletePackingPlan(topologyName));
+    logFine("Set new PackingPlan: %s",
+        stateManager.setPackingPlan(proposedProtoPackingPlan, topologyName));
+
+    // delete the physical plan so TMaster doesn't try to re-establish it on start-up.
+    logFine("Deleted Physical Plan: %s", stateManager.deletePhysicalPlan(topologyName));
+
+    if (removableContainerCount > 0) {
+      scalableScheduler.get().removeContainers(containerDelta.getContainersToRemove());
+    }
+  }
+
+  @VisibleForTesting
+  TopologyAPI.Topology getUpdatedTopology(String topologyName,
+                                          PackingPlan proposedPackingPlan,
+                                          SchedulerStateManagerAdaptor stateManager) {
+    TopologyAPI.Topology updatedTopology = stateManager.getTopology(topologyName);
+    Map<String, Integer> proposedComponentCounts = proposedPackingPlan.getComponentCounts();
+    return mergeTopology(updatedTopology, proposedComponentCounts);
+  }
+
+  @VisibleForTesting
+  void validateCurrentPackingPlan(PackingPlans.PackingPlan existingProtoPackingPlan,
+                                  String topologyName,
+                                  SchedulerStateManagerAdaptor stateManager) {
+    PackingPlans.PackingPlan foundPackingPlan = stateManager.getPackingPlan(topologyName);
+    assertTrue(foundPackingPlan.equals(existingProtoPackingPlan),
+        "Existing packing plan received does not equal the packing plan found in the state "
+            + "manager. Not updating updatedTopology. Received: %s, Found: %s",
+        existingProtoPackingPlan, foundPackingPlan);
+  }
+
+  @VisibleForTesting
+  static class ContainerDelta {
+    private final Set<PackingPlan.ContainerPlan> containersToAdd;
+    private final Set<PackingPlan.ContainerPlan> containersToRemove;
+
+    @VisibleForTesting
+    ContainerDelta(Set<PackingPlan.ContainerPlan> currentContainers,
+                   Set<PackingPlan.ContainerPlan> proposedContainers) {
+      Set<PackingPlan.ContainerPlan> toAdd = new HashSet<>();
+      for (PackingPlan.ContainerPlan proposedContainerPlan : proposedContainers) {
+        if (!currentContainers.contains(proposedContainerPlan)) {
+          toAdd.add(proposedContainerPlan);
+        }
+      }
+      this.containersToAdd = Collections.unmodifiableSet(toAdd);
+
+      Set<PackingPlan.ContainerPlan> toRemove = new HashSet<>();
+      for (PackingPlan.ContainerPlan curentContainerPlan : currentContainers) {
+        if (!proposedContainers.contains(curentContainerPlan)) {
+          toRemove.add(curentContainerPlan);
+        }
+      }
+      this.containersToRemove = Collections.unmodifiableSet(toRemove);
```

Comment:
I'm a little confused here. Do we remove a container if its instances do not exactly match the previous ones? What happens if the current container contains instances 1,2,3 and the proposed container contains 1,2,4? Do we remove this container and re-allocate it?